pkg/ingester/client/compat.go

```diff
@@ -276,11 +282,57 @@ func (s Sample) MarshalJSON() ([]byte, error) {
 func (s *Sample) UnmarshalJSON(b []byte) error {
 	var t model.Time
 	var v model.SampleValue
-	vs := [...]json.Unmarshaler{&t, &v}
+	vs := [...]stdjson.Unmarshaler{&t, &v}
 	if err := json.Unmarshal(b, &vs); err != nil {
 		return err
 	}
 	s.TimestampMs = int64(t)
 	s.Value = float64(v)
 	return nil
 }
+
+func init() {
+
+	jsoniter.RegisterTypeEncoderFunc("client.Sample", func(ptr unsafe.Pointer, stream *jsoniter.Stream) {
+		sample := (*Sample)(ptr)
+
+		stream.WriteArrayStart()
+		stream.WriteRaw(model.Time(sample.TimestampMs).String())
+		stream.WriteMore()
+		stream.WriteString(model.SampleValue(sample.Value).String())
+		stream.WriteArrayEnd()
+	}, func(unsafe.Pointer) bool {
+		return false
+	})
+
+	jsoniter.RegisterTypeDecoderFunc("client.Sample", func(ptr unsafe.Pointer, iter *jsoniter.Iterator) {
+		if !iter.ReadArray() {
+			iter.ReportError("client.Sample", "expected [")
+			return
+		}
+
+		t := model.Time(iter.ReadFloat64() * float64(time.Second/time.Millisecond))
+
+		if !iter.ReadArray() {
+			iter.ReportError("client.Sample", "expected ,")
+			return
+		}
+
+		bs := iter.ReadStringAsSlice()
+		ss := *(*string)(unsafe.Pointer(&bs))
+		v, err := strconv.ParseFloat(ss, 64)
```

Comment:
Why can't we use `iter.ReadFloat64()` again?

Reply:
The Prometheus JSON encodes the sample values as a string containing a number (due to JSON spec numbers not having enough precision I think).  So we need to read a string and parse it ourself.

Reply:
Perhaps inability to represent NaN?

Reply:
Yeah, could be.